lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -1102,17 +1108,41 @@ static std::optional<int> dotCanBeProperlyAsync(ttng::DotAsyncOp dotOp,
     return std::nullopt;
   }
 
-  // Rule 2: The dot should only be used by the for loop's `yield`.
-  if (!dotOp->hasOneUse() ||
-      *dotOp->getUsers().begin() != forOp.getBody()->getTerminator()) {
-    LDBG("Can't make dot async because it is not used only by the loop's "
-         "`yield`.");
-    return std::nullopt;
+  // Rule 2: The dot cannot be unconditionally used by any op in the loop.
+  // Uses under `if` are allowed, as can be explicitly synced with a `wait 0`.
+  int iterArgIdx = -1;
+  Value iterArg = nullptr;
+  SmallVector<std::pair<Operation *, int>> queue;
+  for (auto &use : dotOp->getUses()) {
+    queue.push_back({use.getOwner(), use.getOperandNumber()});
+  }
+  while (!queue.empty()) {
+    auto [user, argIdx] = queue.pop_back_val();
+    if (user->getParentOp() == forOp) {
+      if (isa<scf::YieldOp>(user)) {
+        if (iterArg) {
+          // The dot is used by the loop's yield, but we can't have any other
+          // uses.
+          return std::nullopt;
+        }
+        iterArgIdx = argIdx;
+        iterArg = forOp.getRegionIterArg(argIdx);
+        continue;
+      }
+      return std::nullopt;
+    }
+    if (auto ifOp = dyn_cast<scf::IfOp>(user->getParentOp())) {
+      if (isa<scf::YieldOp>(user)) {
+        // The result is returned by the if, follow it further.
+        auto uses = ifOp.getResult(argIdx).getUses();
+        for (auto &use : uses) {
+          queue.push_back({use.getOwner(), use.getOperandNumber()});
+        }
+      }
+    } else {
```

Comment:
do we really need the recursing logic? How about we check that the uses are only yield or are within a IF block. Then we insert a wait 0 in the IF block. The result should be always correct

Reply:
We also need to find if the dot result is used in the loop yield transitively through the `if`, to check bunch of conditions for the next iteration (the code below my change). And `if`s can be nested, so recursive logic. Can I find this in a simpler way?

Reply:
yeah you're totally right, I missed that we needed that case.